bitbots_vision/src/bitbots_vision/vision_modules/yolo_handler.py

```diff
@@ -262,6 +267,291 @@ def predict(self):
+                if scale < threshold:
+                    continue
+                box_index = self._entry_index(params.side, params.coords, params.classes, n * side_square + i, 0)
+                # Network produces location predictions in absolute coordinates of feature maps.
+                # Scale it to relative coordinates.
+                x = (col + predictions[box_index + 0 * side_square]) / params.side
+                y = (row + predictions[box_index + 1 * side_square]) / params.side
+                # Value for exp is very big number in some cases so following construction is using here
+                try:
+                    w_exp = exp(predictions[box_index + 2 * side_square])
+                    h_exp = exp(predictions[box_index + 3 * side_square])
+                except OverflowError:
+                    continue
+                # Depending on topology we need to normalize sizes by feature maps (up to YOLOv3) or by input shape (YOLOv3)
+                w = w_exp * params.anchors[2 * n] / (resized_image_w if params.isYoloV3 else params.side)
+                h = h_exp * params.anchors[2 * n + 1] / (resized_image_h if params.isYoloV3 else params.side)
+                for j in range(params.classes):
+                    class_index = self._entry_index(params.side, params.coords, params.classes, n * side_square + i,
+                                            params.coords + 1 + j)
+                    confidence = scale * predictions[class_index]
+                    if confidence < threshold:
+                        continue
+                    objects.append(self._scale_bbox(x=x, y=y, h=h, w=w, class_id=j, confidence=confidence,
+                                            h_scale=orig_im_h, w_scale=orig_im_w))
+        return objects
+
+    def _intersection_over_union(self, box_1, box_2):
+        """
+        Calculates an iou of two bounding boxes.
+
+        :param box_1: first bounding box
+        :param box_2: second bounding box
+        :return: iou value
+        """
+        # Calculate overlap
+        width_of_overlap_area = min(box_1['xmax'], box_2['xmax']) - max(box_1['xmin'], box_2['xmin'])
+        height_of_overlap_area = min(box_1['ymax'], box_2['ymax']) - max(box_1['ymin'], box_2['ymin'])
+        if width_of_overlap_area < 0 or height_of_overlap_area < 0:
+            area_of_overlap = 0
+        else:
+            area_of_overlap = width_of_overlap_area * height_of_overlap_area
+        # Calculate areas
+        box_1_area = (box_1['ymax'] - box_1['ymin']) * (box_1['xmax'] - box_1['xmin'])
+        box_2_area = (box_2['ymax'] - box_2['ymin']) * (box_2['xmax'] - box_2['xmin'])
+        # Calculate union
+        area_of_union = box_1_area + box_2_area - area_of_overlap
+        # Catch divide by zero
+        if area_of_union == 0:
+            return 0
+        # Return iou
+        return area_of_overlap / area_of_union
+
+    def _scaled_bbox_to_candidate(self, bbox):
```

Comment:
Isn't the bbox of input just a any bbox? So it is only scaled, if a already scaled bbox is given?

Reply:
Explain this further in person.